commit c87d3120859bbd4f983d9f1305a898eac765f42f
Author: Michael Still <mikal@stillhq.com>
Date:   Wed May 13 11:37:01 2026 +1000

    Allow requiring secure channels with SPICE
    
    The nova_spice_require_secure configuration option requires clients
    to connect to SPICE consoles using channels secured by TLS. The
    default is `no` to match the previous behaviour of allowing insecure
    channels.
    
    Note that nova_spice_require_secure implies that qemu is configured
    to support TLS, and that TLS certificates have been configured in
    the nova_libvirt container.
    
    Partial-Bug: #2131793 (spice-direct consoles)
    Change-Id: I4306725fb8433860baec8e01c51e9eae2beb3cc9
    Signed-off-by: Michael Still <mikal@stillhq.com>

diff --git a/ansible/roles/nova-cell/defaults/main.yml b/ansible/roles/nova-cell/defaults/main.yml
index 790190fb0..b82dde11c 100644
--- a/ansible/roles/nova-cell/defaults/main.yml
+++ b/ansible/roles/nova-cell/defaults/main.yml
@@ -688,6 +688,11 @@ nova_cell_database_enable_tls_internal: "{{ database_enable_tls_internal | bool
 
 nova_multi_compute_ironic_config: []
 
+##################
+# Spice options
+##################
+nova_spice_require_secure: false
+
 ###################
 # Copy certificates
 ###################
diff --git a/ansible/roles/nova-cell/tasks/precheck.yml b/ansible/roles/nova-cell/tasks/precheck.yml
index 9364c619e..b61d2292c 100644
--- a/ansible/roles/nova-cell/tasks/precheck.yml
+++ b/ansible/roles/nova-cell/tasks/precheck.yml
@@ -146,3 +146,14 @@
     - container_facts.containers['nova_libvirt'] is defined
     - not nova_libvirt.enabled | bool
     - inventory_hostname in groups[nova_libvirt.group]
+
+- name: Checking that libvirt_tls is enabled when nova_spice_require_secure is true
+  ansible.builtin.assert:
+    msg: >-
+      nova_spice_require_secure requires libvirt_tls to be enabled, as the
+      SPICE TLS configuration reuses the libvirt TLS certificates.
+    that:
+      - libvirt_tls | bool
+  when:
+    - inventory_hostname in groups[nova_cell_compute_group]
+    - nova_spice_require_secure | bool
diff --git a/ansible/roles/nova-cell/templates/nova-libvirt.json.j2 b/ansible/roles/nova-cell/templates/nova-libvirt.json.j2
index e81413b35..f1118adf4 100644
--- a/ansible/roles/nova-cell/templates/nova-libvirt.json.j2
+++ b/ansible/roles/nova-cell/templates/nova-libvirt.json.j2
@@ -74,5 +74,29 @@
             "owner": "root",
             "perm": "0600"
         }{% endif %}
+{#- The SPICE TLS bind-mounts below are owned root:nova with mode 0440
+    because qemu reads these files directly when starting the SPICE
+    server, and qemu runs as user 'nova' per qemu.conf. The libvirt_tls
+    certs above use 0600 owned root because libvirtd (running as root)
+    is the consumer for those, not qemu. -#}
+{% if nova_spice_require_secure | bool %},
+        {
+            "source": "{{ container_config_directory }}/serverkey.pem",
+            "dest": "/etc/pki/libvirt-spice/server-key.pem",
+            "owner": "root:nova",
+            "perm": "0440"
+        },
+        {
+            "source": "{{ container_config_directory }}/servercert.pem",
+            "dest": "/etc/pki/libvirt-spice/server-cert.pem",
+            "owner": "root:nova",
+            "perm": "0440"
+        },
+        {
+            "source": "{{ container_config_directory }}/cacert.pem",
+            "dest": "/etc/pki/libvirt-spice/ca-cert.pem",
+            "owner": "root:nova",
+            "perm": "0440"
+        }{% endif %}
     ]
 }
diff --git a/ansible/roles/nova-cell/templates/nova.conf.j2 b/ansible/roles/nova-cell/templates/nova.conf.j2
index 7d30dcfd0..7dbbffc8f 100644
--- a/ansible/roles/nova-cell/templates/nova.conf.j2
+++ b/ansible/roles/nova-cell/templates/nova.conf.j2
@@ -79,6 +79,7 @@ novncproxy_base_url = {{ nova_novncproxy_fqdn | kolla_url(public_protocol,  nova
 enabled = false
 [spice]
 enabled = true
+require_secure = {{ nova_spice_require_secure | bool }}
 server_listen = {{ api_interface_address }}
 image_compression = {{ nova_spice_image_compression }}
 jpeg_compression = {{ nova_spice_jpeg_compression }}
diff --git a/ansible/roles/nova-cell/templates/qemu.conf.j2 b/ansible/roles/nova-cell/templates/qemu.conf.j2
index 03b56479a..fb0010783 100644
--- a/ansible/roles/nova-cell/templates/qemu.conf.j2
+++ b/ansible/roles/nova-cell/templates/qemu.conf.j2
@@ -6,6 +6,11 @@ group = "nova"
 max_files =  {{ qemu_max_files }}
 max_processes =  {{ qemu_max_processes }}
 
+{% if nova_spice_require_secure | bool %}
+spice_tls = 1
+spice_tls_x509_cert_dir = "/etc/pki/libvirt-spice"
+{% endif %}
+
 {% if enable_nova_swtpm | bool %}
 swtpm_user = "{{ swtpm_user }}"
 swtpm_group = "{{ swtpm_group }}"
diff --git a/doc/source/reference/compute/nova-guide.rst b/doc/source/reference/compute/nova-guide.rst
index 1069030bb..9ac7d6f08 100644
--- a/doc/source/reference/compute/nova-guide.rst
+++ b/doc/source/reference/compute/nova-guide.rst
@@ -49,6 +49,14 @@ serial console support can be enabled by setting
 
 * nova_spice_html5: configures whether the HTML5 transcoding proxy used by
   Horizon is enabled, and defaults to ``true``.
+* nova_spice_require_secure: configures whether SPICE connections require that
+  the connection be secured by TLS. This defaults to ``false`` to remain
+  compatible with previous installations. Note that ``true`` implies
+  configuration and deployment of TLS certificates.
+
+Additionally ``spice`` has the following configuration options passed directly
+through to ``nova``:
+
 * nova_spice_image_compression: which compression algorithm to use for images
   when using the SPICE console type. Defaults to ``auto_glz``.
 * nova_spice_jpeg_compression: whether or not to use JPEG compression with
diff --git a/releasenotes/notes/expose-spice-secure-6fb03214fa8ecc75.yaml b/releasenotes/notes/expose-spice-secure-6fb03214fa8ecc75.yaml
new file mode 100644
index 000000000..97273a86d
--- /dev/null
+++ b/releasenotes/notes/expose-spice-secure-6fb03214fa8ecc75.yaml
@@ -0,0 +1,9 @@
+---
+other:
+  - |
+    This release exposes additional configuration options for OpenStack
+    Compute consoles which utilize the SPICE protocol. You can now require
+    that such consoles use channels secured by TLS using the
+    ``nova_spice_require_secure`` configuration option. This option defaults
+    to ``false`` as it implies that TLS certificates have been distributed
+    to your ``nova-compute`` nodes.
